daal4py/sklearn/_utils.py

```diff
@@ -257,3 +266,131 @@ def get_status(self, logs=False):
+
+    @wraps(original_init)
+    def init_with_n_jobs(self, *args, n_jobs=None, **kwargs):
+        original_init(self, *args, **kwargs)
+        self.n_jobs = n_jobs
+
+    # add "n_jobs" parameter to signature of wrapped init
+    # if estimator doesn't originally support it
+    sig = signature(original_init)
+    original_params = list(sig.parameters.values())
+    if "n_jobs" not in list(map(lambda param: param.name, original_params)):
+        original_params.append(Parameter("n_jobs", Parameter.KEYWORD_ONLY, default=None))
+        init_with_n_jobs.__signature__ = sig.replace(parameters=original_params)
+
+    original_class.__init__ = init_with_n_jobs
+
+    # add n_jobs to __doc__ string if needed
+    if (
+        hasattr(original_class, "__doc__")
+        and isinstance(original_class.__doc__, str)
+        and "n_jobs : int" not in original_class.__doc__
+    ):
+        parameters_doc_tail = "\n    Attributes"
+        n_jobs_doc = """
+    n_jobs : int, default=None
+        The number of jobs to use in parallel for the computation.
+        ``None`` means using all physical cores
+        unless in a :obj:`joblib.parallel_backend` context.
+        ``-1`` means using all logical cores.
+        See :term:`Glossary <n_jobs>` for more details.
+"""
+        original_class.__doc__ = original_class.__doc__.replace(
+            parameters_doc_tail, n_jobs_doc + parameters_doc_tail
+        )
+
+    return original_class
+
+
+# Note: getting controller in global scope of this module is required
+# to avoid overheads by its initialization per each function call
+threadpool_controller = threadpoolctl.ThreadpoolController()
+
+
+def get_suggested_n_threads(n_cpus):
+    """Function to get `n_threads` limit
+    if `n_jobs` is set in upper parallelization context.
+    Usually, limit is equal to `n_logical_cpus` // `n_jobs`.
+    Returns None if limit is not set.
+    """
+    n_threads_map = {
+        lib_ctl.internal_api: lib_ctl.get_num_threads()
+        for lib_ctl in threadpool_controller.lib_controllers
+        if lib_ctl.internal_api != "mkl"
+    }
+    # openBLAS is limited by 128 threads by default.
+    # thus, 128 threads from openBLAS is uninformative
+    if "openblas" in n_threads_map and n_threads_map["openblas"] == 128:
+        del n_threads_map["openblas"]
+    # remove default values equal to n_cpus as uninformative
+    for backend in list(n_threads_map.keys()):
+        if n_threads_map[backend] == n_cpus:
+            del n_threads_map[backend]
+    if len(n_threads_map) > 0:
+        return min(n_threads_map.values())
+    else:
+        return None
+
+
+def run_with_n_jobs(method):
+    """Decorator for running of methods containing oneDAL kernels with 'n_jobs'"""
+
+    @wraps(method)
+    def method_wrapper(self, *args, **kwargs):
+        # threading parallel backend branch
+        if not isinstance(threading.current_thread(), threading._MainThread):
```

Comment:
Possibly a dumb question: In the case that someone spawns a sklearnex routine in additional thread, it this will just run as normal?

Reply:
It will just run as normal as it was before this change. Another issue is an oversubscription which might occur by calling oneDAL in different threads.